irods/test/data_obj_test.py

```diff
@@ -2094,6 +2105,132 @@ def test_touch_operation_does_not_work_when_given_a_collection__525(self):
         with self.assertRaises(ex.InvalidInputArgument):
             user_session.data_objects.touch(home_collection_path)
 
+    @unittest.skipIf(progressbar is None, "progressbar is not installed")
+    def test_pbar_for_parallel_io_4(self):
+        from irods.manager.data_object_manager import (register_update_instance, _update_fns)
+
+        class wrapper:
+            def __init__(self,pbar):
+                self.pbar = pbar
+                self.lock = threading.Lock()
+                self.total = 0
+
+            def update(self,n):
+                with self.lock:
+                    self.total += n
+                    self.pbar.update(self.total)
+
+        LEN = 1024**2*40
+        content = b'_'*LEN
+
+        progress_bar = progressbar.ProgressBar(maxval=len(content))
+        wrapped = wrapper(progress_bar)
+
+        register_update_instance(wrapped, wrapped.update)
+
+        wrapped.pbar.start()
+        self._run_pbars_for_parallel_io(content, [wrapped])
+        self.assertEqual(wrapped.pbar.currval, LEN)
+
+    @unittest.skipIf(progressbar is None, "progressbar is not installed")
+    def test_pbar_for_parallel_io_3(self):
+        from irods.manager.data_object_manager import (register_update_type, unregister_update_type)
+
+        class ProgressBar_wrapper:
+         def start(self,*x):return self.p.start(*x)
+         def update(self,*x):return self.p.update(*x)
+         @property
+         def currval(self): return self.p.currval
+         def __init__(self,*args,**kw):
+          self.p = progressbar.ProgressBar(*args,**kw)
+
+        def adapt_ProgressBar(pbar, total = weakref.WeakKeyDictionary(), # Any mutable parameters thus introduced to the function scope
+                                    l = threading.Lock()):               # are one-time initialized & memoized for subsequent calls.
```

Comment:
Can you expand on the comments here? Maybe even as a docstring? I've run into problems when using default values like this, so I'm curious about how it works.

Reply:
Sure can.  Yep, it seems there were some uses for which mutable default values were perfectly, and secretly intended.

Reply:
A little revision and extra review here. Since the commits added yesterday, the factory functions are going to be called only once and the result now properly cached, so that in all cases, each instance gets  the factory called  on it only once.  Thus where we were using a lot of these mutable default args, we now are no longer. We're using closures instead.  Since the change, the only way one would depend on such a one-time generated default argument would be to protect an entire progress bar library ( not just each instance) against multithreaded access.

Which lets you know it's probably a mistake to depend on those default values heavily, in an interface design : )

Reply:
The thing to remember is, the default argument value is generated only once, and that's when the function declaration is encountered. This only matters, of course, when it's mutable (such as a threading.Lock object is) because developers may gloss over and not catch the fact that the default value could change over multiple invocations of the function object.  Thus the memoization effect if you send a mutable (eg a list), in other words the ability of a function to detect how many times, or even when(!) it was last called!

Reply:
Okay. Maybe that explanation (or something like it) in a docstring could be helpful for future readers?

Reply:
New comments seem better, yes